File: core/data/src/main/java/org/mifos/mobilewallet/core/domain/usecase/account/DownloadTransactionReceipt.java
package org.mifos.mobilewallet.core.domain.usecase.account;

import org.mifos.mobilewallet.core.base.UseCase;
import org.mifos.mobilewallet.core.data.fineract.repository.FineractRepository;
import org.mifos.mobilewallet.core.utils.Constants;

import javax.inject.Inject;

import okhttp3.ResponseBody;
import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * Created by ankur on 06/June/2018
 */

public class DownloadTransactionReceipt extends
        UseCase<DownloadTransactionReceipt.RequestValues,
                DownloadTransactionReceipt.ResponseValue> {

    private final FineractRepository mFineractRepository;

    @Inject
    public DownloadTransactionReceipt(FineractRepository fineractRepository) {
        mFineractRepository = fineractRepository;
    }

    @Override
    protected void executeUseCase(RequestValues requestValues) {

        mFineractRepository.getTransactionReceipt(Constants.PDF, requestValues.transactionId)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(new Subscriber<ResponseBody>() {
                    @Override
                    public void onCompleted() {

                    }

                    @Override
                    public void onError(Throwable e) {
                        getUseCaseCallback().onError(e.toString());
                    }

                    @Override
                    public void onNext(ResponseBody responseBody) {
                        getUseCaseCallback().onSuccess(new ResponseValue(responseBody));
                    }
                });
    }

    public static final class RequestValues implements UseCase.RequestValues {

        private final String transactionId;

        public RequestValues(String transactionId) {
            this.transactionId = transactionId;
        }
    }

    public static final class ResponseValue implements UseCase.ResponseValue {

        private final ResponseBody responseBody;

        public ResponseValue(ResponseBody responseBody) {
            this.responseBody = responseBody;
        }

        public ResponseBody getResponseBody() {
            return responseBody;
        }
    }
}

File: core/data/src/main/java/org/mifospay/core/data/domain/usecase/account/DownloadTransactionReceipt.kt
/*
 * Copyright 2024 Mifos Initiative
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * See https://github.com/openMF/mobile-wallet/blob/master/LICENSE.md
 */
package org.mifospay.core.data.domain.usecase.account

import okhttp3.ResponseBody
import org.mifospay.core.data.base.UseCase
import org.mifospay.core.data.fineract.repository.FineractRepository
import org.mifospay.core.data.util.Constants
import rx.Subscriber
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers

class DownloadTransactionReceipt (
    private val mFineractRepository: FineractRepository,
) : UseCase<DownloadTransactionReceipt.RequestValues, DownloadTransactionReceipt.ResponseValue>() {
    override fun executeUseCase(requestValues: RequestValues) {
        requestValues.transactionId?.let {
            mFineractRepository.getTransactionReceipt(Constants.PDF, it)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(
                    object : Subscriber<ResponseBody>() {
                        override fun onCompleted() {}
                        override fun onError(e: Throwable) {
                            useCaseCallback.onError(e.toString())
                        }

                        override fun onNext(t: ResponseBody) {
                            useCaseCallback.onSuccess(ResponseValue(t))
                        }
                    },
                )
        }
    }

    data class RequestValues(val transactionId: String?) : UseCase.RequestValues
    data class ResponseValue(val responseBody: ResponseBody) : UseCase.ResponseValue
}
